### PR TITLE
:memo: Add typing to some callback classes

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
Provides an exemple to #16691

### CODE REVIEW

File: keras/callbacks.py [hunk 1]
```diff
@@ -1318,15 +1327,17 @@ class ModelCheckpoint(Callback):
...
+        filepath: Union[str, os.PathLike],
+        monitor: str = "val_loss",
+        verbose: int = 0,
+        save_best_only: bool = False,
+        save_weights_only: bool = False,
+        mode: str = "auto",
+        save_freq: Union[int, str] = "epoch",
+        options: Union[
+            tf.train.CheckpointOptions, tf.saved_model.SaveOptions, None
+        ] = None,
```
Comment (fchollet) [inline keras/callbacks.py:1339]: In general we should avoid explicit types when the expect type is a custom class, like in the above case. It creates significant tech debt and maintenance burden going forward.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! First, we'll discuss with the team to determine whether we need/want explicit types throughout the codebase.

**Comment by fchollet** [comment]
After extensive discussions, this is what we will do going forward:

> We accept type annotations for simple Python types (not unions, and not custom types), but they're optional when adding new code. We will merge contributor PRs that add such annotations going forward. We will not maintain custom types.

Please update the PR accordingly and format the code using `black`. Thanks!

**Comment by gabrieldemarmiesse** [comment]
Could we make an exception for the public api? I get that it's a lot of work to use custom types everywhere in the codebase and to maintain it, but for the public API, it has a direct benefit to the user through Mypy and IDEs, so it might outweight the cost of maintenance

Furthermore, those public API type hints could be in theory be read and displayed in the documentation, giving us even more incentive to do it

**Comment by fchollet** [comment]
This is our policy at this time. We may revise it in the future.

**Review by fchollet — APPROVED** [review]
Thanks for the updates!

**Pull request merged**